### ISSUE TITLE
Remove native modules optional dependency

### ISSUE THREAD

**Issue by ehmicky** [brightcove/hot-shots#178]
Some users might want to use this library but without UDS support. For those, installing a native module might be prohibitive. Native modules can make `npm install` much slower. Also their compilation can fail. For example, installing `hot-shots` on my machine (Node 14.7.0) results in the bug described [here](https://github.com/bnoordhuis/node-unix-dgram/issues/58):

```
> unix-dgram@2.0.4 install /home/ether/Desktop/node_modules/unix-dgram
> node-gyp rebuild

make: Entering directory '/home/ether/Desktop/node_modules/unix-dgram/build'
  CXX(target) Release/obj.target/unix_dgram/src/unix_dgram.o
In file included from ../src/unix_dgram.cc:5:
../../nan/nan.h: In function ‘void Nan::AsyncQueueWorker(Nan::AsyncWorker*)’:
../../nan/nan.h:2294:62: warning: cast between incompatible function types from ‘void (*)(uv_work_t*)’ {aka ‘void (*)(uv_work_s*)’} to ‘uv_after_work_cb’ {aka ‘void (*)(uv_work_s*, int)’} [-Wcast-function-type]
 2294 |     , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
      |                                                              ^
In file included from ../../nan/nan.h:56,
                 from ../src/unix_dgram.cc:5:
../src/unix_dgram.cc: At global scope:
/home/ether/.cache/node-gyp/14.7.0/include/node/node.h:746:43: warning: cast between incompatible function types from ‘void (*)(v8::Local<v8::Object>)’ to ‘node::addon_register_func’ {aka ‘void (*)(v8::Local<v8::Object>, v8::Local<v8::Value>, void*)’} [-Wcast-function-type]
  746 |       (node::addon_register_func) (regfunc),                          \
      |                                           ^
/home/ether/.cache/node-gyp/14.7.0/include/node/node.h:780:3: note: in expansion of macro ‘NODE_MODULE_X’
  780 |   NODE_MODULE_X(modname, regfunc, NULL, 0)  // NOLINT (readability/null_usage)
      |   ^~~~~~~~~~~~~
../src/unix_dgram.cc:404:1: note: in expansion of macro ‘NODE_MODULE’
  404 | NODE_MODULE(unix_dgram, Initialize)
      | ^~~~~~~~~~~
  SOLINK_MODULE(target) Release/obj.target/unix_dgram.node
  COPY Release/unix_dgram.node
make: Leaving directory '/home/ether/Desktop/node_modules/unix-dgram/build'
```

While there are solutions to the problem above, native modules tend to create problems like this. 

Making them an `optionalDependencies` does not solve this. It only makes `npm install` succeed when the compilation fails, but it does not remove the time spent compiling. Unfortunately, npm does not have a way to disable specific `optionalDependencies`, it only has a way to disable all of them.

One way to provide UDS without forcing users into installing native module would be to create a second separate package with UDS support, which would use the same core logic as this package. What do you think?

**Comment by bdeitte** [comment]
Thanks for all the info @ehmicky, appreciate it!  I see in the linked issue that there is a solution now- does that change how you feel about this?  I'm wondering if the overhead of having a full separate package would be worth it for the one optional install here.  I'm not sure, would want to hear if there's more people who feel this is needed.

**Comment by ehmicky** [comment]
Hi @bdeitte,
Thanks for your answer! Unfortunately not. At the moment, I am using an alternative statsd client, but I'd prefer to use `hot-shots` because it has more dogstatsd-specific logic

**Comment by bdeitte** [comment]
Hi @ehmicky given I haven't heard of others needing this and overhead on keeping a separate package, I am going to close this out.  If there's a way that does seem simple to solve this in the future, certainly good with having something here.

**Comment by matej-prokop** [comment]
Hey @bdeitte I know that it is probably fault of npm that it doesn't provide way to exclude (opt-out) some optional dependency. However, by adding `unix-dgram` as optional dependency to `hot-shots` project, it effectively forced us to stay at 6.2.0 (last version without `unix-dgram`). Tens of our projects have `hot-shots` as transitive dependency and upgrading hot-shots would "break" their build. In our use-case we don't need/use UDS but `npm` doesn't give you any way to opt-out.

Even if you release once a year, version without UDS as dependency that would be awesome. I know it is ugly but I don't know any other solution and I would really love to keep using `hots-shots` :(

**Comment by bdeitte** [comment]
Curious how this broke your build @matej-prokop?  Is it that you're not allowed to use unix-dgram, unix-drgram won't install properly, something else?  There may be other solutions here so curious.

**Comment by matej-prokop** [comment]
Sorry - now I see that my previous message for probably missleading.

During `npm install` npm always tries to install `unix-dgram` but that fails because gyp is missing in our environment. Entire `npm install` finishes because `unix-dgram` is optional dependency so `npm` ignores/skips this failure. So for instance build in CI is passing and everything is nicely green.       

Nevertheless, actual problem is perception from user perspective (for user that did `npm install`). That "failing" installation of `unix-dgram` produces huge error message into output (see attached screenshot below); therefore, user thinks that there is something really wrong or that it just failed completely. They ping you `"there is something wrong with your package"`, you reply with `"please ignore that error that's fine"` and they say `"ignoring errors is way to hell, please fix it"` and you `"sorry I can't"`. Now, imagine there are plenty of such users so rolling that out to all of them would cause avalanche of complains and pushback. 

![image](https://user-images.githubusercontent.com/7286793/100845890-24a13380-347e-11eb-98ad-bffc233b086d.png)

**Comment by bdeitte** [comment]
Got it, thanks for the info.  Well I don't think having a different version here is still going to be the solution unfortunately.  You could try something like this?  https://stackoverflow.com/questions/33542712/npm-never-install-nested-optional-dependencies-for-npm-package

**Comment by mariusandra** [comment]
I'll second that this optional dependency is making me look for an alternative statsd client (for [posthog-plugin-server](https://github.com/PostHog/posthog-plugin-server)). Our logs are cluttered with this gyp failure and it's hard to parse them and see that actually everything is fine. 

![image](https://user-images.githubusercontent.com/53387/102402789-92ed0680-3fe5-11eb-98cb-9a13b8591cbc.png)

A clean approach out of this would be to pass the `const socket = require('unix-dgram').createSocket().bind('/path')` variable to hot-shots during init. For example as the `hostname`. That way whoever wants can require and use it outside...

**Comment by mariusandra** [comment]
I published [@posthog/hot-shots](https://www.npmjs.com/package/@posthog/hot-shots) version 8.2.1. It's exactly the same package, just the `optionalDependencies` line is removed... and thus no unix-dgram support anyway.

This package will probably stay up, but nobody should really use on it for their app. I don't plan on keeping it in sync with the main repo and we will probably remove it once this issue is resolved nicely...

**Comment by matej-prokop** [comment]
> You could try something like this? https://stackoverflow.com/questions/33542712/npm-never-install-nested-optional-dependencies-for-npm-package

@bdeitte Thank you for suggestion! I was already experimenting with shrinkwrap because of `unix-dgram` but it doesn't work in desired way. Even if you do `npm install hot-shots --no-optional` and `npm shrinkwrap` it writes `unix-dgram` into `npm-shrinkwrap.json` as depedency. Therefore running just `npm install` later results in an attempt to install `unix-dgram` 😞 

![image](https://user-images.githubusercontent.com/7286793/102595117-65a07580-4117-11eb-9499-52b6837b0a8c.png)

**Comment by amakhrov** [comment]
Sad to see this closed. We're having a 30s overhead on each job that installs NPM modules in CI (which we have multiple in our complex pipelines) for compiling `unix-dgram` even though we don't really use this prototocl
